### PR TITLE
Replace espree and doctrine with typescript in docgen.

### LINKED ISSUE #17271: Documentation: Improve the way types are presented in the generated documentation
Follow-up for #17014.

Related comment from @sirreal https://github.com/WordPress/gutenberg/pull/17014/files#r317982877:

> I find the `[string]` syntax strange. Where does it come from?
> 
> I assume it means "optional parameter with a default value", but that only seems to vaguely align with jsdoc, where `[square brackets]` are used on the parameter or property names to indicate they're optional, not around the type.
> 
> To add to the confusion, in TypeScript `[ string ]` would mean a one-tuple of strings, or an array with a single `string` member, e.g. `[ "foo" ]` but never `[]` or `[ "foo", "bar" ]`.

Response from @dsifford:

> Yeah, as it stands, the way that the JSDoc is being parsed definitely results in clunky markdown for this particular case.
> 
> In JSDoc, it's pretty easy to tell the difference between optional types and single-tuple types because they are written differently.
> 
> ```js
> // optional string type
> /**
>  * @param {string} [foo]
>  */
> 
> // single tuple string type
> /**
>  * @param {[string]} foo
>  */
> ```
> 
> So I think that you've definitely raised a valid concern with respect to how the docs are being generated. 
> 
> (PS: syntax comes from [here](https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html#param-and-returns))

### LINKED ISSUE #18045: docgen: remove doctrine
`doctrine` is the JSDoc parser we use in `docgen` to get structured information out of the JSDoc comments. It was created and maintained by the ESLint people, although it's [no longer actively developed](https://eslint.org/blog/2018/11/jsdoc-end-of-life). We should migrate out of `doctrine` for that reason.

The suggested replacement ([eslint-plugin-jsoc](https://github.com/gajus/eslint-plugin-jsdoc)) uses a different [JSDoc parser](https://github.com/jsdoctypeparser/jsdoctypeparser) with support for many other syntaxes (typescript, etc). We should look into trying `jsdoctypeparser` or any other that's actively developed.

### LINKED ISSUE #19952: docgen: Support ES2019, approved ES2020 syntax
Previously: https://github.com/WordPress/gutenberg/issues/18045#issuecomment-578161548
Related: https://github.com/WordPress/gutenberg/pull/19931#issuecomment-579847333, #19831

Docgen uses a different JavaScript parser ([Espree](https://github.com/eslint/espree)) than that which is used in the plugin build process ([Babel](https://babeljs.io/)). This hasn't been much of an issue thus far, but with especially new syntaxes, it can cause errors to be thrown when running `npm run docs:build` and with the corresponding pre-commit step.

Examples:

- [ES2019 optional catch binding](https://2ality.com/2017/08/optional-catch-binding.html)
   - Was intended to be used in #19871
- [ES2020 optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining)
   - Was intended to be used in #19931

Originally, I thought it might just be a matter of keeping the `espree` dependency up to date, assuming that it's been updated for newer versions of the specification.

This is unfortunately not as simple as it might seem, since one of the breaking changes introduced in [`espree@5.0.0`](https://github.com/eslint/espree/releases/tag/v5.0.0) is to remove the `attachComment` feature, which is pretty critical to how `docgen` works (or at least is currently implemented). Thus, upgrading to the latest version would require a pretty significant refactor to the code, one where we'd likely emulate how [comment attachment](https://github.com/eslint/espree/pull/405/files#diff-474d681dc53859768e80de42f3d9a4ca) worked previously, using the [`comment` (and perhaps `tokens`) option(s)](https://github.com/eslint/espree#usage) to infer where a comment occurs in relation to an exported member. I started down this path on Friday, but it seems significant enough to warrant whether it's worth the effort, considering if the alternative to Doctrine would bring with it a different parser implementation as well (e.g. TypeScript).

One thing I noted in the process of this is that we configure the ECMAScript version here:

https://github.com/WordPress/gutenberg/blob/1fb3c386001266ad03ed6d2cd3d9ad8777090028/packages/docgen/src/engine.js#L15

This is particularly relevant for my [previous comment](https://github.com/WordPress/gutenberg/issues/18045#issuecomment-578161548) where I mention being unable to use ES2019 features. It's pretty clear by this code why this wouldn't be expected to work 😄 I haven't tested yet, but it's possible that Espree versions prior to the breaking 5.0 version might have support for newer versions of ECMAScript, so that we could still use those language features without bumping to the latest Espree version.

**Task:** Allow new syntaxes to be parsed without error by `docgen`.

**Proposed solution:**

One of:

1. Replace Doctrine and Espree with TypeScript (#18045).
2. Or: Bump Espree to the latest version, including necessary refactoring to `docgen` to support removal of `attachComment` option.
3. Or: Optionally bump Espree to the last version including `attachComment` option, and check to see whether it can support ES2019 and ES2020 as the configured `ecmaVersion` option.

**Follow-up tasks:**

Revisit the "intended to be used" examples from above, to introduce the desired syntax.

- [ ] https://github.com/WordPress/gutenberg/pull/19871/files#r372503756
- [ ] https://github.com/WordPress/gutenberg/pull/19931#issuecomment-579847333

### PR DESCRIPTION
**Pull request by sainthkh**
* Closes #19952
* Closes #18045 
* Closes #17271

## Description
* Replace espree and doctrine with typescript: parsing the code and generating docgen IR code is done by typescript.
  - Now, we can print types like `(hello: string) => string`.
* Used ES2019, ES2020 syntax
  - #19871: optional catch binding
  - #19931: optional chaining

## How has this been tested?
* Tested by refactoring docgen tests. 
* And added tests when necessary. (Like type generation)

## Screenshots 

N/A

## Types of changes

Bug fix

## Changes

* `Object` -> `object`. TypeScript treats `Object` as `any`. And it's the first `don't` of the [do's and don'ts list](https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#number-string-boolean-symbol-and-object). It might need a lot of changes in the code. 
  - In TypeScript, `object` means `{}` thing. `Object` means [this global object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object). That's why it is not recommended to use `Object`.
* `Array<X>`, `Array.<X>`, `X[]` -> `X[]`. Type generator doesn't respect how it is written. Actually, there is no way to find which is written in which way. I thought `[]` is a visual way of telling "it's an array"
  - The alternative might use `[]` for primitives and `Array<>` for user types that start with a capital letter like `Array<ReactNode>` or all `Array<>`.
* I removed line numbers in IR nodes. Because we need to calculate it every time we need it. It is not calculated by default like espree 4.x
* TypeScript notation over JSDoc. I did it because it's more clear. It means:
  - `string | undefined` over `string=`
  - `number | nullable over `?number`
* When there is a default value, it's added at the back like this: `-   _status_ `string`: Notice status. (Default: `info`)` [check here](https://github.com/WordPress/gutenberg/commit/c72d628e85daad29e3a1b7bc5e2445540cb37054#diff-5fe8f8fafe931b38ddfc0e2e2ef483cdR71)
  - I thought the type name is a bit vague and thought about adding `(optional)` after name like `boolean (optional)`. I'm not sure about this and didn't implement it. 
* I indented properties like below. It's how TypeScript handles jsdoc objects and I think it looks better.
```md
-   _options_ `object | undefined`: Notice options.
    -   _context_ `string`: Context under which to group notice. (Default: `global`)
    -   _id_ `string | undefined`: Identifier for notice. Automatically assigned if not specified.
    -   _isDismissible_ `boolean`: Whether the notice can be dismissed by user. (Default: `true`)
    -   _type_ `string`: Type of notice, one of `default`, or `snackbar`. (Default: `default`)
    -   _speak_ `boolean`: Whether the notice content should be announced to screen readers. (Default: `true`)
    -   _actions_ `WPNoticeAction[] | undefined`: User actions to be presented with notice.
```
[Check code here](https://github.com/WordPress/gutenberg/commit/c72d628e85daad29e3a1b7bc5e2445540cb37054#diff-5fe8f8fafe931b38ddfc0e2e2ef483cdR73-R79)

## Later works

In this PR, I focused on migrating and tried not to add new features as much as possible. Because it's big (changes in 120 files with +3,454 −8,408) and it can make code hard to review and get merged. 

The list below is the things that should be added for better developer experience.

* Create separate eslint jsdoc config for projects type-checked by typescript.
  - TypeScript checks types, we don't need some `eslint-plugin-jsdoc` configs like `valid-types`. We can remove `/* eslint-disable */`s for perfectly ok types like `{[setting:string]:any}`.
  - Turn off '`object` -> `Object`' rule.
* Handle JSDoc types defined in non-last or non-exported comments. This makes users check source code to see what's going on.
  - Currently, `docgen` cannot handle docgen types like below:
```js
/**
 * Type below is ignored.
 * @typedef WPLinkControlProps
 *
 * @property {(WPLinkControlSetting[])=} settings An array of settings objects.
 */

/**
 * Function
 * @param {WPLinkControlProps} props Component props.
 */
export function LinkControl( {
	value,
} ) { }
```
```js
/**
 * Type below is ignored.
 * @typedef WPLinkControlProps
 *
 * @property {(WPLinkControlSetting[])=} settings An array of settings objects.
 */

// Value for test.
const CREATE_TYPE = '__CREATE__';

/**
 * Function
 * @param {WPLinkControlProps} props Component props.
 */
export function LinkControl( {
	value,
} ) { }
```
* Throw errors when something wrong happened like `wrong type name`, `Object`.
* Known issues for parsing default values
  - array in objects like `[p={x:[1, 2]}]`.
  - escaping string-opening character like `"abc\"def\"ghi"`. 
* Add docgen to TS-enabled project. (#18838)
* Updating TS for docgen isn't just changing numbers. For example, we need to add new [`awaited` type](https://github.com/microsoft/TypeScript/pull/35998) for TS3.9
* Debug mode: print out ast and ir to files. It's not possible now because of the `parent` element of the TypeScript ast.

## Note
Before this PR is merged, it should be rebased and docs should be regenerated. If not, it can break other PRs.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [N/A] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [N/A] I've included developer documentation if appropriate. 
- [N/A] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by nosolosw** [comment]
Hey, @sainthkh thanks for working on this, and I'm excited to help with this. I want to set expectations as to when would I be able to, though: I think I can come to this by next week.

**Comment by sainthkh** [comment]
@nosolosw Thanks. I can wait. There're some PRs of mine that took 3+ months to merge. It doesn't matter to me.

**Comment by gziolo** [comment]
@sainthkh, just wanted to say that I appreciate all the work spent on this PR 👍 

@nosolosw is the best person to validate the implementation proposed.

@aduth and @sirreal – they should be able to help validate all the changes proposed to bridge JSDoc and TypeScript.

File: packages/docgen/scripts/dump-ast.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+const ts = require( 'typescript' );
+const { readFileSync } = require( 'fs' );
+const { inspect } = require( 'util' );
+
+const filename = process.argv[ 2 ];
+
+const sourceFile = ts.createSourceFile(
+	filename,
+	readFileSync( filename ).toString(),
+	ts.ScriptTarget.ES2020
```
Comment (sirreal) [inline packages/docgen/scripts/dump-ast.js]: Should we use an evergreen target like `ESNext`?

```suggestion
	ts.ScriptTarget.ESNext
```

File: packages/docgen/src/get-export-statements.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * hasExportModifier
+ *   - export const x = 42
+ *   - export default class {}
+ *   - export default class ClassDeclaration {}
+ *   - export default function() {}
+ *   - export default function myDeclaration() {}
+ *   - export class MyDeclaration {}
+ *   - export function myDeclaration() {}
+ *
```
Comment (sirreal) [inline packages/docgen/src/get-export-statements.js]: Some cases that don't seem to be covered in this list, are these handled properly?

```js
export { default as foo, createContext as cc } from 'react'
const bar = 'baz'
const myDefault = 'd'
export { bar as quux, myDefault as default }
```

File: packages/docgen/src/get-jsdoc-from-statement.js [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+
+		return parseFloat( rawValue );
+	}
+
+	if ( rawValue === 'null' ) return null;
+	if ( rawValue === 'true' ) return true;
+	if ( rawValue === 'false' ) return false;
+
+	if ( rawValue[ 0 ] === '{' && rawValue[ rawValue.length - 1 ] === '}' ) {
+	}
```
Comment (sirreal) [inline packages/docgen/src/get-jsdoc-from-statement.js]: Empty conditional block here?

File: packages/docgen/src/get-jsdoc-from-statement.js [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+		if ( Number.isInteger( rawValue ) ) {
+			return parseInt( rawValue );
+		}
+
+		return parseFloat( rawValue );
+	}
+
+	if ( rawValue === 'null' ) return null;
+	if ( rawValue === 'true' ) return true;
+	if ( rawValue === 'false' ) return false;
```
Comment (sirreal) [inline packages/docgen/src/get-jsdoc-from-statement.js]: I didn't realize we allowed this, I'd expect linter to enforce braces on these conditional blocks.

File: packages/docgen/src/index.js [hunk 5]
```diff
@@ -126,14 +77,14 @@ module.exports = function( sourceFile, options ) {
...
-		process.exit( 0 );
-	}
+	// if ( result === undefined ) {
+	// 	process.stdout.write(
+	// 		'\nFile was processed, but contained no ES6 module exports:'
+	// 	);
+	// 	process.stdout.write( `\n${ sourceFilePath }` );
+	// 	process.stdout.write( '\n\n' );
+	// 	process.exit( 0 );
+	// }
```
Comment (sirreal) [inline packages/docgen/src/index.js]: Remove or restore this commented out section?

**Review by sirreal — COMMENTED** [review]
This is an amazing effort we'll really benefit from. Thanks!

I've done an initial scan and left some notes and questions.

File: packages/block-editor/src/components/link-control/index.js [hunk 6]
```diff
@@ -210,8 +210,7 @@ function LinkControl( {
 		// is still unassigned after calling `onChange`.
 		const hadFocusLoss =
 			isEndingEditWithFocus.current &&
-			wrapperNode.current &&
-			! wrapperNode.current.contains( document.activeElement );
+			! wrapperNode.current?.contains( document.activeElement );
```
[the remark this answers is not in the document]
Comment (sainthkh) [reply packages/block-editor/src/components/link-control/index.js]: This line of code is why everything started. After #19831, everyone thought that we could use optional chaining in Gutenberg. 

But we couldn't because `espree` 4.x in docgen cannot parse it. @aduth thought we can simply do it by updating `espree` to the latest version. Unfortunately, espree doesn't support `leadingComment` properties after 5.0. So, we had to change the parser. (Older version of docgen extract JSDoc from `leadingComment` and parsed it with `doctrine`. And `doctrine` is deprecated and unused.)

As TypeScript can parse JS and JSDoc, it was chosen. 

This line means that we can use optional chaining without any error.

File: packages/docgen/src/compile.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * Function that takes file path and returns compiled result
+ *
+ * @param {string} filePath the path to the file to be compiled
+ */
+module.exports = function( filePath ) {
+	const options = {
+		allowJs: true,
+		target: ts.ScriptTarget.ES2020,
```
[the remark this answers is not in the document]
Comment (sainthkh) [reply packages/docgen/src/compile.js]: It isn't important because this option is for the **transpiled result** of the code. It's not for how TypeScript treats new syntaxes. In other words, even when we set this to `ES3`, TypeScript can compile optional chaining and other ES2015-2020 features. But in this case, the result of the transpilation can be run on the environments that only support ES3 (maybe notorious IE9?).

We don't need transpiled JavaScript in docgen. All we need is the generated AST. So, it doesn't matter. 

But to avoid confusion, I removed target option from `compile.js` and changed `ES2020` to `ESNext` in other files.

File: packages/docgen/src/compile.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+	};
+	const host = ts.createCompilerHost( options, true );
+	const raw = readFileSync( filePath ).toString();
+	const code = raw
+		// typescript interprets @wordpress in @example code as a JSDoc tag.
+		// So, it should be replaced for the time being.
+		.replace( /@wordpress/g, '__WORDPRESS_IMPORT__' )
```
[the remark this answers is not in the document]
Comment (sainthkh) [reply packages/docgen/src/compile.js]: 1\. I commented when they're restored. 

2\. That can be a problem when we have written code that exports imported items directly from other packages like this:

```js
// packages/some-package/src/index.js

import escapeHtml from '@wordpress/escape-html';

export { escapeHtml }
```

In this case, the parser tries to open `packages/escape-html/src/index.js` and find jsdoc of `escapeHtml`. I guess it's more of a design flaw than a feature. Don't you think so?

File: packages/docgen/src/get-export-entries.js [hunk 9]
```diff
@@ -1,103 +1,116 @@
...
+	if ( hasExportModifier( statement ) ) {
+		if ( hasDefaultModifier( statement ) ) {
+			const getLocalName = ( s ) => {
+				switch ( s.kind ) {
+					case SyntaxKind.ClassDeclaration:
+					case SyntaxKind.FunctionDeclaration:
+					default:
+						return s.name ? s.name.text : '*default*';
+				}
+			};
```
[the remark this answers is not in the document]
Comment (sainthkh) [reply packages/docgen/src/get-export-entries.js]: When I was migrating getLocalName(), I copied the original function first to think what can be changed. 

[In the original](https://github.com/WordPress/gutenberg/blob/master/packages/docgen/src/get-export-entries.js#L24-L37), `switch` makes perfect sense. But in the new one, we don't need it. So, I removed it entirely.

File: packages/docgen/src/get-export-statements.js (hunk 2 above)
Comment (sainthkh) [reply packages/docgen/src/get-export-statements.js]: I added them here because I want to be clear which condition handles which AST case. I thought your examples could be deduced from `export { a, b } from 'c'` and `export { myDeclaration };`, but it seems that it wasn't. I added them. 

And they're handled properly and tested with unit tests.

File: packages/docgen/src/get-jsdoc-from-statement.js (hunk 3 above)
Comment (sainthkh) [reply packages/docgen/src/get-jsdoc-from-statement.js]: I first thought about parsing objects and arrays. But I realized that to add them to docs, I need to return them back to strings. And it's unnecessary. So, I gave it up. 

But I forgot to remove the code for that attempt.

File: packages/docgen/src/get-jsdoc-from-statement.js (hunk 4 above)
Comment (sainthkh) [reply packages/docgen/src/get-jsdoc-from-statement.js]: For some reason, [`curly` rule](https://eslint.org/docs/rules/curly) doesn't work. When I add it to root `.eslintrc.js`, it works. But somehow, this rule is ignored although it's set to [ 'error', 'all' ] in `eslint-plugin/configs/es5.js`.

Fixed them anyway. 

We need to open an issue for this, right?

File: packages/docgen/src/index.js (hunk 5 above)
Comment (sainthkh) [reply packages/docgen/src/index.js]: Revived it.

**Comment by sainthkh** [comment]
@sirreal Thanks for all the reviews. I fixed them all.

**Comment by sainthkh** [comment]
Flaky failure.

File: packages/block-editor/src/components/link-control/index.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (aduth) [reply packages/block-editor/src/components/link-control/index.js:213]: I don't think we need to back it out at this point, but I'd generally suggest to find ways to limit the scope of a pull request as much as possible, especially if it's already a +3400/-8400 proposal. It's definitely related as the original motivation of the changes, but I think it could have been fine to do as a separate follow-up pull request.

File: packages/docgen/scripts/dump-ast.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+const filename = process.argv[ 2 ];
+
+const sourceFile = ts.createSourceFile(
+	filename,
+	readFileSync( filename ).toString(),
+	ts.ScriptTarget.ESNext
+);
+
+// eslint-disable-next-line no-console
+console.log( inspect( sourceFile, false, null ) );
```
Comment (aduth) [inline packages/docgen/scripts/dump-ast.js:17]: Two minor points:

- We should want to document our inline disablings (related: #16795)
- We could probably avoid it if we used `process.stdout.write` instead, which also might be the more "correct" way to arbitrarily write to stdout (?)

File: packages/docgen/src/compile.js [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+	const raw = readFileSync( filePath ).toString();
+	const code = raw
+		// typescript interprets @wordpress in @example code as a JSDoc tag.
+		// So, it should be replaced for the time being.
+		// They're restored in get-jsdoc-from-statement.js
+		.replace( /@wordpress/g, '__WORDPRESS_IMPORT__' )
+		// When <caption>ES(5|Next)<\/caption> exists next to @example tag,
+		// typescript cannot parse code correctly.
+		// So, they're removed.
+		.replace( /@example\s+<caption>ES(5|Next)<\/caption>/g, '@example' );
```
Comment (aduth) [inline packages/docgen/src/compile.js:25]: It'd probably be fine to address separately, but our usage of HTML after the `@example` tag feels weird. I wonder if it would help avoid this logic if we at least moved those `<caption>` tags onto the line immediately following `@example` ?

File: packages/docgen/src/compile.js [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+	const sF = ts.createSourceFile( filePath, code, options.target, true );
```
Comment (aduth) [inline packages/docgen/src/compile.js:26]: I think it's reasonably self-explanatory based on the function name from which it's being assigned, but in general I would always encourage naming variables in their full, non-abbreviated form (i.e. `sourceFile`), since abbreviations can cause unnecessary difficulty in interpreting for future maintenance. We can afford the few extra characters 😄

**Review by aduth — COMMENTED** [review]
@sainthkh This is a huge undertaking, and I applaud your willingness in taking it on! I left only a few comments for today and barely scratched the surface. I think I'll need to return to this another time when I can dedicate some deep thought.

No worries if there's not, but: Is there anything we can remove from this pull request to make it smaller? Or tasks we can do separately which would eliminate some of the accommodations we needed to make here ([example](https://github.com/WordPress/gutenberg/pull/21238#discussion_r402611495))? I understand it'll be a big set of changes regardless, but if we can pull out individual tasks to tackle individually, it would go a long way to helping this along.

File: packages/docgen/src/compile.js (hunk 12 above)
Comment (sainthkh) [reply packages/docgen/src/compile.js:26]: I named it in that way, because there was a `sourceFile` below to be exported. I haven't read the TypeScript code, but I guess that they're same. 

But to make it clear, I changed it from `sF` to `internalSourceFile`.

File: packages/docgen/src/compile.js (hunk 11 above)
Comment (sainthkh) [reply packages/docgen/src/compile.js:25]: Yes, it does. [Check this out](https://astexplorer.net/#/gist/959d69eaf758ea2281438e5a2b510bba/972255808da8964e869afea19e795e17ca1ce0cc) 

But I think we should add it to the 'later works' list. To fix it, we need to change 10 more files.

File: packages/docgen/scripts/dump-ast.js (hunk 10 above)
Comment (sainthkh) [reply packages/docgen/scripts/dump-ast.js:17]: I changed it to `process.stdout.write`.

File: packages/block-editor/src/components/link-control/index.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (sainthkh) [reply packages/block-editor/src/components/link-control/index.js:213]: I reverted it and added new test `new-syntax` under `test/fixtures`. Later, when new syntax doesn't work or want to check if a new syntax works, we can use this file.

**Comment by sainthkh** [comment]
@aduth

1\. When I was writing this, I also thought about the length of this PR. In my experience, when PR is too long, they're abandoned because it's too hard to follow the changes. 

2\. But the problem is that this PR is based on the entirely new AST. As TypeScript should incorporate its own type syntaxes, they had to create their own AST, so TypeScript AST is quite different from other ASTs. 

(When you compare the ASTs of these 3 different parses ([espree](https://astexplorer.net/#/gist/0ae9707edba3f2605a88537dca8df336/3ad0c3d9605b883fbaeb2bf9d953df99f3eb6446), [babel](https://astexplorer.net/#/gist/4aad28dad128a87b86e70e10a49b43a4/c0cc4757fd7e28fd82863727cffbe28d96a70cc3), [recast](https://astexplorer.net/#/gist/089e4793aaa75c6c1a5dd1ef7e14ae80/6325b8a6fd52a9e5bf54de7431f2e20c88d08b89)), they're mostly same.)

Because of that, almost all of docgen code is re-written. 

That's why it's hard to break this PR into a few smaller PRs. One step is tightly related to the next step.

3\. That made me also think how to make this PR easier to review.

First thing I did was to revert `Update Doc` commit for the time being. It'll break `Build artifacts` test.  [I tested it with my local clone.](https://github.com/sainthkh/gutenberg/pull/2/files) It became much easier to review. We can update docs when code review is done.

I'm also thinking about writing about the code flow and other helpful notes or even setting up small goals to focus on small areas of code. But I also think this can bias the code review. I'm a bit hesitant about setting goals for review, or should I?

I also thought about removing changes in the list of changes above. It does make us discuss less by postponing some issues later in other new issues. But it doesn't make me remove some codes to make it easier to review.

File: packages/docgen/src/compile.js (hunk 11 above)
Comment (aduth) [reply packages/docgen/src/compile.js:25]: > But I think we should add it to the 'later works' list. To fix it, we need to change another 10 more files.

What I find can work well in these situations is: Create a separate pull request which includes those smaller independent changes. At the time it's merged, the branch of this pull request can be rebased to account for those new revisions. For this case, I expect it means we wouldn't need this code anymore to handle this syntax.

**Comment by sainthkh** [comment]
I created 2 mini PRs: #21404, #21405.

**Comment by gziolo** [comment]
I landed https://github.com/WordPress/gutenberg/pull/21405 👍

**Comment by nosolosw** [comment]
Hi @sainthkh, thanks for your patience! The first thing I want to say is that I really appreciate all the love and work you've put into this, bravo! :clap: 

This is promising, although I was wondering what do you think of these:

- One of the issues we have now is that we're using two different parsers (espree for docs & babel for code). If we do this, we'll still have two, so I worry that we may end up in the same place we're now. Is there any way we can consolidate both? Can we use `babel` for docgen and delegate the JSDoc (with TypeScript notation) to something else (perhaps the `typescript` parser)?

- From what I gather, the TypeScript AST is different from the ESTree-based ones that most tools in the ecosystem use. I was thinking whether this could pose problems when/if we want to integrate with some of those (ex: reusing parts of [react-docgen](https://github.com/reactjs/react-docgen) to document the [largely undocumented](https://github.com/WordPress/gutenberg/issues/14227) React components we have, testing, etc).

**Comment by sainthkh** [comment]
1\.

When I first decided to use `typescript`, I expected TypeScript to correctly parse and generate type name from TypeScript AST. And as we all know, it didn't. (Because of that, I needed to create [this file](https://github.com/WordPress/gutenberg/blob/0ec03a53dfac26ba0a74e2049d83026c76a78e21/packages/docgen/src/test/fixtures/tags-param-types/code.js).) Major reason for choosing TypeScript was false. So, I don't mind changing it to `@babel/parser`. When we use babel, we can reuse most of the code. (I think it's another advantage.)

2\.

I am against using `typescript` as a JSDoc parser. Because when you run the typescript-version of docgen command. You can feel that it's slower than espree version. It seems that typescript has a lot of things to load. (It's not a surprise because it needs to load its type checkers beside JS parsing.)

There is a package called [`comment-parser`](https://www.npmjs.com/package/comment-parser). If we decide to use `@babel/parser` as the main parser for JS, I would use it as the parser for JSDoc comments. 

It's a simple parser. It becomes both our advantage and disadvantage.

* advantage: It doesn't throw an error if the type is invalid. (Our type generator and VS Code can handle this problem.) It's easy to implement the types. It parses default value out of the box (TypeScript doesn't.)
* disadvantage: It cannot normalize type names. `X[]`, `Array<X>`, `Array.<X>` are all same types. But it prints out all of them as they're written. We need to discuss whether it's right to normalize them, if yes, how to normalize them.

Maybe, we can try [@babel/plugin-transfrom-typescript](https://babeljs.io/docs/en/babel-plugin-transform-typescript) to parse the types, but I don't know it works. It needs experiment. 

3\.

As I don't mind changing to Babel, the second question is automatically solved.

**Comment by nosolosw** [comment]
@sainthkh thanks for the additional context!

I was thinking about your comment and the next steps:

1. We change `espree` by `@babel/parser` in a separate PR (keep this PR one, just in case). This will fix the syntax issues that we now have.
2. We look at the alternatives we have to change doctrine after, in a new PR.

It feels to me that way we're fixing the most pressing issues (using new syntax) and we buy some time to look into the TypeScript integration. However, to be honest, I don't know what kind of TypeScript issues we may run into in step 2 as I'm still leveling up in that area (although I've looked at the logs and it doesn't seem to me that we have now any blockers). I'd welcome a sanity check, though @sainthkh @aduth @sirreal do you think this is sound?

**Comment by sainthkh** [comment]
Good enough. I'll test `babel typescript` to check if we can reuse type generator code in this PR.

**Comment by sainthkh** [comment]
@nosolosw 

I was reading Gutenberg code to implement #21807. It made me react a bit slow. Now, I finished replacing `espree` with `babel` in #21853.

----

I experimented if babel can generate type ast. And it can. [Here's the proof.](https://astexplorer.net/#/gist/5ace7b984ada5a78812c30a59283a678/edfb78ddaa4e84b5c8b6f1f7b82b12fd855f9e0e). All we need to do is add `"typescript"` option to the `plugins`. 

We can parse types with babel by generating code like `type X = string | number` and parse it. 

----

### Todo list

- [ ] Replace espree with babel (#21807)
- [ ] Use optional chaining and optional catch binding
- [ ] Replace doctrine with comment-parser
- [ ] Add new type generator

**Comment by aduth** [comment]
Will we still want this pull request in its current form, given that we've merged #21853? I understand there are still tasks remaining, but it's unclear to me if we'll pursue those as separate independent pull requests, or if you'd prefer to try to reconcile the changes proposed in this branch to support them.

**Comment by aduth** [comment]
> * Replace doctrine with comment-parser

Noting that I remarked some comments on initial experimenting to use `comment-parser` (including a patch) at https://github.com/WordPress/gutenberg/issues/18045#issuecomment-559916571 . Overall it didn't work as well as I expected, though it may still end up being the best tool we have available, without going to the extent of creating something totally new.

**Comment by sainthkh** [comment]
* I left this PR open to cherry-pick the related codes. 
* I'm currently experimenting with `comment-parser`. Maybe we can parse and verify type definition with `@babel/parser` and `jsdoctypeparser`. After experiment is done, I'll share you with the result.

**Comment by gziolo** [comment]
> I left this PR open to cherry-pick the related codes

@sainthkh, any plans to reuse more code from this PR or can it be closed?
